design/external-references.md

```diff
@@ -0,0 +1,394 @@
+
+**Synthesis-time Imports**: we will also define an environmental context provider that will import values during synthesis and store them in `cdk.json`. This approach is more robust since it will allow CDK code to reason about concrete values instead of opaque tokens. It's especially needed for situations where a list of values needs to be passed and the arity is required.
+
+**Convenience methods**: now that constructs can be serializable through imports/exports, we can implement a set of convenience methods for each AWS resource to provide nice ergonomics for the specific use case of exporting and importing resources across apps.
+
+## Design
+
+### Serialization
+
+If a construct is said to be serializable, it must implement the `ISerializable` interface:
+
+```ts
+interface ISerializable {
+  serialize(ctx: ISerializationContext): void;
+}
+```
+
+When an object is serialized, the `serialize` method is called with an object which implements the following interface:
+
+```ts
+interface ISerializationContext {
+  writeString(key: string, value: string, options?: SerializationOptions): void
+  writeObject(key: string, obj: ISerializable, options?: SerializationOptions): void;
+}
+
+interface SerializationOptions {
+  description?: string;
+}
+```
+
+The serialization context allows the object serialize itself through key/value strings via calls to `writeString(key, value)`. If the object encapsulates another serializable object, it can use `writeObject(key, obj)`, which will result in a subsequent call to the sub-object's `serialize` method with an appropriate context.
+
+### Deserialization
+
+To support deserialization, classes must also include a public static `deserializeXxx` method which reads the object from a deserialization context and returns an object that implements the resource type interface:
+
+> The reason we indicate the type in the method name is because static methods in JavaScript are inherited, so we can differentiate between `ExtraBucket.deserializeExtraBucket` and `ExtraBucket.deserizlizeBucket`.
+
+```ts
+class MyResource extends Construct implements ISerializable {
+  static deserializeMyResource(ctx: IDeserializationContext): IMyResource;
+}
+```
+
+The deserialization context is an object that implements the following interface:
+
+```ts
+interface IDeserializationContext {
+  scope: Construct;
+  id: string;
+  readString(key: string, options?: ReadStringOptions): string;
+  readObject(key: string): IDeserializationContext;
+}
+
+interface ReadStringOptions {
+  allowUnresolved?: boolean; /** @default true */
+}
+```
+
+`readString(key)` can be used to read values stored by `writeString`. and `readObject(key)` returns a deserialization context for composite deserialization written via `writeObject`.
```

Comment:
It'd be nice to have a systematic/standard/recommended way to handle object shape changes. Schema versioning of some kind.

Also - this does not seem to cater for required/optional. Example: *SecretsManager* Secrets can have a *KMS* key or not. Same goes for buckets...